Adds Gitlab Nulecule example

navidshaikh opened the pull request:
Fixes: https://github.com/projectatomic/nulecule/issues/160
- Added Gitlab Nulecule example for Docker and Kubernetes provider
- Added answer.conf.sample for reference

navidshaikh committed:
Adds Kubernetes artifacts for gitlab examples

navidshaikh committed:
Adds Dockerfile Nulecule and answer.conf.sample for gitlab eg

navidshaikh committed:
Adds README license for gitlab example

navidshaikh committed:
Updates examples/gitlab-centos7-atomicapp/README.md

navidshaikh committed:
Adds artifacts for Docker provider

 Updates the answer.conf.sample file for Docker provider

navidshaikh committed:
Updates Nulecule file with Docker provider

navidshaikh committed:
Updates the postgresql container Docker artifact

navidshaikh committed:
Adds answers.conf.sample in the atomicapp image

navidshaikh committed:
Updates answers.conf contents gitlab example README

navidshaikh committed:
Fixes a typo and adds note about Docker provider in gitlab example

vpavlin commented on examples/gitlab-centos7-atomicapp/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+
+LABEL io.projectatomic.nulecule.specversion="0.0.2" \
+      io.projectatomic.nulecule.providers="kubernetes"\
+      Build="docker build --rm --tag swordphilic/gitlab-atomicapp ."
```
What about removing your user name from the tag?:-)

navidshaikh replied:
Right! :) 
BTW, is `Build` a recommended LABEL?

vpavlin replied:
We don't have any "official" usage for it, but it's probably nice to have if you have some fancy bash aliases set up (https://github.com/vpavlin/docker-tools/blob/master/bash_aliases_docker#L65)

navidshaikh replied:
This is interesting! I will pull these aliases, thanks! :-)

vpavlin commented on examples/gitlab-centos7-atomicapp/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+      io.projectatomic.nulecule.providers="kubernetes"\
+      Build="docker build --rm --tag swordphilic/gitlab-atomicapp ."
+
+ADD /Nulecule /Dockerfile README.md gpl-3.0.txt answers.conf.sample /application-entity/
```
`answers.conf.sample` is generated on `atomic install`, no need to package it

navidshaikh replied:
I wanted to provide a sample `answer.conf.sample` so that user can view and relate, however I could find the same `answer.conf.sample` after `install` - it generates a minimal answer file with provider and namespace fields in general section. is this expected? how can one package the `answer.conf.sample` file so that after `install` atomicapp can populate it for me in $pwd ?

vpavlin replied:
If you add `params` section with all available parameters they will appear in `answers.conf.sample` with default values after `atomic install`:-)

navidshaikh replied:
ah.. thanks for the tip! :-)

vpavlin commented:
There are no `params` in any of the components although you are using them in `answers` and `artifacts`. It works because atomic app tries to be as nice to user as possible, but it's not a good practice.

Please add params to all components - I see you have $image everywhere for example. See other examples for reference: f.e. https://github.com/projectatomic/nulecule/blob/master/examples/mariadb-centos7-atomicapp/Nulecule#L11

navidshaikh committed:
Adds required parameters in Nulecule for gitlab example

 Updates Nulecule with required parameters, incoporating the review comments by
 vpavlin

navidshaikh committed:
Excludes answer.conf.sample from gitlab atomicapp image

 Also updates the `Build` LABEL
 Incorporates the review comments by vpavlin

navidshaikh commented:
@vpavlin : ACK! thank you for the pointers. I have incorporated the review comments. Kindly review.

vpavlin commented on examples/gitlab-centos7-atomicapp/Nulecule:
```diff
@@ -0,0 +1,63 @@
+        default: gitlab_production
+      - name: NODE_PORT
+        description: Node port where Gitlab HTTP service will be forwarded to (For Kubernetes provider valid range is 30000-32667)
+        default: 30000
```
I'd suggest to use [Constraints](https://github.com/projectatomic/nulecule/tree/master/spec#parameters-object) here as we don't promote them much and this is exactly a place where I'd use them

navidshaikh replied:
I see only the `allowed_pattern` for strings. How do I use it for (int) port range constraint ?

vpavlin commented on examples/gitlab-centos7-atomicapp/Nulecule:
```diff
@@ -0,0 +1,63 @@
+        description: PostgreSQL image to be used.
+        default: swordphilic/postgresql:latest
+      - name: DB_USER
+        description: Database user (This parameter is used for Docker provider)
```
Why is the comment about Docker provider here? I see all the params used in Kubernetes manifests also..

navidshaikh replied:
For Docker provider, these parameters are used in postgresql component while for Kubernetes these params are used in gitlab component.

vpavlin replied:
Can you please explain why is that? (Why for one provider you need params in one component and for different provider you don't need them?)

vpavlin replied:
Ah, got it, that due to the `--link`..hmm..the comment about param being specific to a single provider feels weird as the user will always get asked for all params

navidshaikh replied:
The answer file wont be pretty as it will contain for different providers mixing their flavors! ;)

vpavlin replied:
`answers.conf.sample` will always contain all params no matter what provider you select..

vpavlin commented:
@navidshaikh I'd suggest dropping answers.conf.sample as it's auto generated on `atomic install` call

vpavlin commented on examples/gitlab-centos7-atomicapp/README.md:
```diff
@@ -0,0 +1,157 @@
+
+    $ [sudo] atomic run projectatomic/gitlab-centos7-atomicapp
+
+## Option 2: Unattended
```
IMO _unattended_ should consist of 

```
atomic install ...
mv answers.conf.sample answers.conf
vi answers.conf
atomic run ...
```

We shouldn't suggest any copy&pasting if it's not necessary - get rid of copy&pasting stuff from github is exactly one of the reasons why Nulecule has been created.

This is a basic workflow. There is also an _interactive_ run where you are prompted for missing param values if there is no default. But what you have as _Install and Run_ below is actually the unattended workflow.

@aweiteka @goern comments?

navidshaikh replied:
@vpavlin : The Unattended section follows https://github.com/projectatomic/nulecule#option-2-unattended. And as u said, _Install and Run_ and _Unattended_ methods will be same, if followed the install->rename->update->run approach.

goern replied:
Right, Option 2 sounds like 'configure deployment' rather than unattended, by creating an answers.conf file (and I would get rid of the opportunity to copy and paste from github... ) the user configures the deployment (run) he is going to do.

goern commented on examples/gitlab-centos7-atomicapp/Nulecule:
```diff
@@ -0,0 +1,63 @@
+  appversion: 1.2.0
+  description: Gitlab.
+graph:
+  - name: redis
```
There is a redis Atomic App, please reuse: https://github.com/navidshaikh/nulecule/blob/fix-160/examples/redis-centos7-atomicapp/Nulecule

navidshaikh replied:
Nulecule supports using graphs as well as defining multiple components. I agree on the part that using graphs is recommended. IMO, we should have the example set showing different ways supported by Nulecule. This example showcase using different tiers by enlisting the respective artifacts and not using any graphs.

vpavlin replied:
I agree with @goern here - showing people that they shouldn't reuse, but rather re-create is not best practice that we want to suggest.

goern commented on examples/gitlab-centos7-atomicapp/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+MAINTAINER Navid Shaikh <nshaikh@redhat.com>
+
+LABEL io.projectatomic.nulecule.specversion="0.0.2" \
+      io.projectatomic.nulecule.providers="kubernetes"\
```
you have artifacts for docker and kube, so add docker here

goern commented on examples/gitlab-centos7-atomicapp/README.md:
```diff
@@ -0,0 +1,157 @@
+This is an atomic application based on the nulecule specification. Docker and Kubernetes are the supported providers. You'll need to run this from a workstation that has the atomic command and kubectl client that can connect to a kubernetes master.
```
Atomic App or atomicapp, thx :)

goern replied:
What is the 'atomic command', maybe worth inserting a link here

goern commented on examples/gitlab-centos7-atomicapp/README.md:
```diff
@@ -0,0 +1,157 @@
+
+        $ [sudo] atomic run projectatomic/gitlab-centos7-atomicapp
+
+### Option 3: Install and Run
```
unattended default deployment

goern commented on examples/gitlab-centos7-atomicapp/README.md:
````diff
@@ -0,0 +1,157 @@
+
+```
+
+To forward port `30000` of ADB box to host machine port `30000`, add following line your Vagrantfile
````
This should be automated by the ADB guys, could you please open an issue with them?

navidshaikh replied:
https://github.com/projectatomic/adb-atomic-developer-bundle/issues/87

navidshaikh committed:
Enlists docker provider in Dockerfile for Gitlab example

Incorporating review comments by goern

navidshaikh committed:
Adds pointer for atomic command in Gitlab exmaple README

Incorporating the review comments by goern

navidshaikh committed:
Adds constraints for NODE_PORT param of Gitlab example

Incorporating review comments by vpavlin

navidshaikh committed:
Updates the regex for Gitlab eg parameter constraints

navidshaikh commented:
Updated PR w.r.t suggestions and reviews. Is this okay now ?

navidshaikh commented:
Automated build of gitlab example at https://hub.docker.com/r/swordphilic/gitlab-centos7-atomicapp/ - it can be used to test if example works.

vpavlin commented on examples/gitlab-centos7-atomicapp/artifacts/kubernetes/postgres-rc.json:
```diff
@@ -0,0 +1,52 @@
+                        "env": [
+                            {
+                                "name": "DB_USER",
+                                "value": "gitlab"
```
Shouldn't these env vars be also changeable same as in gitlab-rc.json?

navidshaikh committed:
Adds env vars in postgresql replication controller manifest

navidshaikh committed:
Sync with Atomic App 0.1.11 release

navidshaikh commented on examples/gitlab-centos7-atomicapp/artifacts/kubernetes/postgres-rc.json:
```diff
@@ -0,0 +1,52 @@
+                            {
+                                "name": "DB_NAME",
+                                "value": "$DB_NAME"
+                            }
```
Added env vars as suggested by @vpavlin at
https://github.com/projectatomic/nulecule/pull/165#discussion_r40999345

navidshaikh commented on examples/gitlab-centos7-atomicapp/Dockerfile:
```diff
@@ -0,0 +1,10 @@
+FROM projectatomic/atomicapp:0.1.11
```
Updated Dockerfile with Atomic App release 0.1.11

goern commented:
This has been moved to https://github.com/projectatomic/nulecule-library/tree/master/gitlab-centos7-atomicapp
